lib/autocomplete-manager.coffee

```diff
@@ -1,139 +1,140 @@
 
-    fileName = path.basename @editor.getBuffer().getPath()
-    for blacklistGlob in blacklist
-      if minimatch fileName, blacklistGlob
-        return true
+    # Handle events from suggestion list
+    @subscriptions.add(@suggestionList.onDidConfirm(@confirm))
+    @subscriptions.add(@suggestionList.onDidCancel(@hideSuggestionList))
 
-    return false
+  handleCommands: ->
+    # Allow autocomplete to be triggered via keymap
+    @subscriptions.add(atom.commands.add 'atom-text-editor',
+      'autocomplete-plus:activate': @runAutocompletion
+    )
 
+  # Private: Finds suggestions for the current prefix, sets the list items,
+  # positions the overlay and shows it
+  runAutocompletion: =>
+    @hideSuggestionList()
+    return unless @editor?
+    return unless @buffer?
+    return if @currentFileBlacklisted()
+    @originalCursorPosition = @editor.getCursorScreenPosition()
+    return unless @originalCursorPosition?
+    currentScopes = @editor.scopeDescriptorForBufferPosition(@originalCursorPosition)?.scopes
+    return unless currentScopes?
 
-  # Private: Handles editor events
-  handleEvents: ->
-    # Close the overlay when the cursor moved without
-    # changing any text
-    @compositeDisposable.add @editor.onDidChangeCursorPosition(@cursorMoved)
+    options =
+      editor: @editor
+      buffer: @buffer
+      position: @originalCursorPosition
+      scopes: currentScopes
+      prefixOfSelection: @prefixOfSelection(@editor.getLastSelection())
 
-    # Is this the event for switching tabs? Dunno...
-    @compositeDisposable.add @editor.onDidChangeTitle(@cancel)
+    # Iterate over all providers, ask them to build suggestion(s)
+    suggestions = []
+    for provider in @providersForScopes(options.scopes)
+      providerSuggestions = provider?.buildSuggestionsShim(options)
+      continue unless providerSuggestions?.length
 
-  # Public: Registers the given provider
-  #
-  # provider - The {Provider} to register
-  registerProvider: (provider) ->
-    unless _.findWhere(@providers, provider)?
-      @providers.push(provider)
-      @compositeDisposable.add provider if provider.dispose?
+      if provider.exclusive
+        suggestions = providerSuggestions
+        break
+      else
+        suggestions = suggestions.concat(providerSuggestions)
 
-  # Public: Unregisters the given provider
-  #
-  # provider - The {Provider} to unregister
-  unregisterProvider: (provider) ->
-    _.remove(@providers, provider)
-    @compositeDisposable.remove(provider)
+    # No suggestions? Cancel autocompletion.
+    return unless suggestions.length
+    @suggestionList?.changeItems(suggestions)
+    @showSuggestionList()
+
+  providersForScopes: (scopes) =>
+    return [] unless scopes?
+    return [] unless @scopes
+    providers = []
+    for scope in scopes
+      if @scopes[scope]?
+        providers = _.union(providers, @scopes[scope])
+    providers.push(@fuzzyProvider) unless _.size(providers) > 0
```

Comment:
Not sure I understand what's going on with the scope matching here? Can you explain it? Looks like you're matching every individual scope from a scope descriptor against the contents of a hash, but I could be wrong. My instinct would be that you'd have the providers registered with scope *selectors*, then match them against the current scope *descriptor* to determine which providers apply.